src/types/any.rs

```diff
@@ -769,16 +746,40 @@ impl PyAny {
         unsafe { ffi::Py_TYPE(self.as_ptr()) }
     }
 
-    /// Casts `self` to a concrete Python object type.
-    ///
-    /// This can cast only to native Python types, not types implemented in Rust.
+    /// Converts this `PyAny` to a concrete Python type.
+    #[deprecated(since = "0.18.0", note = "use the equivalent .downcast()")]
     pub fn cast_as<'a, D>(&'a self) -> Result<&'a D, PyDowncastError<'_>>
     where
         D: PyTryFrom<'a>,
     {
         <D as PyTryFrom<'_>>::try_from(self)
     }
 
+    /// Converts this `PyAny` to a concrete Python type.
+    ///
+    /// This can cast only to native Python types, not types implemented in Rust.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use pyo3::prelude::*;
+    /// use pyo3::types::{PyAny, PyDict, PyList};
+    ///
+    /// Python::with_gil(|py| {
+    ///     let dict = PyDict::new(py);
+    ///     assert!(dict.is_instance_of::<PyAny>().unwrap());
+    ///     let any: &PyAny = dict.as_ref();
+    ///     assert!(any.downcast::<PyDict>().is_ok());
+    ///     assert!(any.downcast::<PyList>().is_err());
+    /// });
+    /// ```
+    pub fn downcast<T>(&self) -> Result<&T, PyDowncastError<'_>>
+    where
+        for<'py> T: PyTryFrom<'py>,
```

Comment:
Do we actually need the quantifier here? Wouldn't

```rust
pub fn downcast<'py, T>(&'py self) -> Result<&'py T, PyDowncastError<'_>>
where
    T: PyTryFrom<'py>,
```

suffice?

Reply:
That's the signature of `cast_as`. I don't actually know if there's any meaningful difference here - so I didn't want to change `downcast`'s signature needlessly.

Reply:
Agreed.

Reply:
I'm actually not a fan of `PyTryFrom` at all - the `try_from` method it introduces also clashes with the 2021 prelude. I wonder if actually we want

```rust
pub fn downcast<'py, T>(&'py self) -> Result<&'py T, PyDowncastError<'_>>
where
    T: PyTypeInfo
```

or _maybe_

```rust
pub fn downcast<'py, T>(&'py self) -> Result<&'py T, PyDowncastError<'_>>
where
    T: PyNativeType,
    &'py T: FromPyObject<'py>
```

... which might open up the way for us to get rid of `PyTryFrom`? (We'd perhaps want to also have an `unsafe fn downcast_unchecked` too.)

---

On the original topic, yes I agree with correcting the lifetime bound as proposed.